src/manager/checkpoint.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: MIT
+use std::collections::hash_map::RandomState;
+use std::collections::{HashMap, HashSet};
+use std::time::Duration;
+
+use anyhow::Result;
+use cid::Cid;
+use fil_actors_runtime::cbor;
+use futures_util::future::join_all;
+use fvm_shared::address::Address;
+use fvm_shared::clock::ChainEpoch;
+use fvm_shared::MethodNum;
+use ipc_gateway::Checkpoint;
+use ipc_sdk::subnet_id::SubnetID;
+use primitives::TCid;
+use tokio::time::sleep;
+
+use crate::config::Subnet;
+use crate::lotus::client::LotusJsonRPCClient;
+use crate::lotus::message::mpool::MpoolPushMessage;
+use crate::lotus::LotusClient;
+
+const CHAIN_HEAD_REQUEST_PERIOD: Duration = Duration::from_secs(10);
+
+/// Starts the checkpoint manager.
+#[allow(dead_code)]
+pub async fn start(subnets: HashMap<String, Subnet>) {
+    let managed_subnets = subnets_to_manage(subnets);
+
+    let mut futures = Vec::new();
+    for (child, parent) in managed_subnets {
+        futures.push(manage_subnet(child, parent));
+    }
+    join_all(futures).await;
+}
+
+/// This function takes a `HashMap<String, Subnet>` and returns a `Vec` of tuples of the form
+/// `(child_subnet, parent_subnet)`, where `child_subnet` is a subnet that we need to actively
+/// manage checkpoint for. This means that for each `child_subnet` there exists at least one account
+/// for which we need to submit checkpoints on behalf of to `parent_subnet`, which must also be
+/// present in the map.
+fn subnets_to_manage(subnets: HashMap<String, Subnet>) -> Vec<(Subnet, Subnet)> {
+    // First, we remap subnets by SubnetID.
+    let subnets_by_id: HashMap<SubnetID, Subnet> = subnets
+        .values()
+        .map(|s| (s.id.clone(), s.clone()))
+        .collect();
+
+    // Then, we filter for subnets that have at least one account and for which the parent subnet
```

Comment:
Should we also add a `validator` field in the config for child subnets to flag the subnets that an IPC agent is part of as validator? I think this filter is useful, but it may also be the case that an IPC agent is syncing with a parent-child pair where it is not participating as a validator and it doesn't need to run the checkpoint process.

Reply:
I don't know. I think that will add unnecessary burden to the user and potentially make them miss checkpoints as they would have to manually ensure that the config is in sync with the subnet actor state.

Reply:
You are right, I saw afterwards that you are explicitly checking if we hold a validator address in the subnet, so no need for this. I forgot to remove this comment, resolving.